src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs

```diff
@@ -2006,12 +2017,71 @@ class Point
+                new { Source = "Point(3, string { Length: 2 })", Type = "System.Object", ConvertedType = "Point" },
+                new { Source = "3", Type = "System.Object", ConvertedType = "System.Int32" },
+                new { Source = "string { Length: 2 }", Type = "System.Object", ConvertedType = "System.String" },
+                new { Source = "2", Type = "System.Int32", ConvertedType = "System.Int32" },
+                new { Source = "(_, { })", Type = "Point", ConvertedType = "Point" },
+                new { Source = "_", Type = "System.Object", ConvertedType = "System.Object" },
+                new { Source = "{ }", Type = "System.Object", ConvertedType = "System.Object" },
+                new { Source = "Point({ }, { }, { })", Type = "Point", ConvertedType = "Point" },
+                new { Source = "{ }", Type = "?", ConvertedType = "?" },
+                new { Source = "{ }", Type = "?", ConvertedType = "?" },
+                new { Source = "{ }", Type = "?", ConvertedType = "?" },
+                new { Source = "Point(, { })", Type = "Point", ConvertedType = "Point" },
+                new { Source = "", Type = "System.Object", ConvertedType = "System.Object" },
+                new { Source = "{ }", Type = "System.Object", ConvertedType = "System.Object" },
+            };
+            var compilation = CreateCompilation(source, options: TestOptions.ReleaseDll);
+            compilation.VerifyDiagnostics(
+                // (10,24): error CS8400: Pattern missing
+                //         if (p is Point(, { })) { }
+                Diagnostic(ErrorCode.ERR_MissingPattern, ",").WithLocation(10, 24),
+                // (9,18): error CS1501: No overload for method 'Deconstruct' takes 3 arguments
+                //         if (p is Point({ }, { }, { })) { }
+                Diagnostic(ErrorCode.ERR_BadArgCount, "Point({ }, { }, { })").WithArguments("Deconstruct", "3").WithLocation(9, 18),
+                // (9,18): error CS8129: No suitable Deconstruct instance or extension method was found for type 'Point', with 3 out parameters and a void return type.
+                //         if (p is Point({ }, { }, { })) { }
+                Diagnostic(ErrorCode.ERR_MissingDeconstruct, "Point({ }, { }, { })").WithArguments("Point", "3").WithLocation(9, 18)
+                );
+            var tree = compilation.SyntaxTrees[0];
+            var model = compilation.GetSemanticModel(tree);
+            int i = 0;
+            foreach (var pat in tree.GetRoot().DescendantNodesAndSelf().OfType<PatternSyntax>())
+            {
+                var typeInfo = model.GetTypeInfo(pat);
+                var ex = expected[i++];
+                Assert.Equal(ex.Source, pat.ToString());
+                Assert.Equal(ex.Type, typeInfo.Type.ToTestDisplayString());
+                Assert.Equal(ex.ConvertedType, typeInfo.ConvertedType.ToTestDisplayString());
             }
+            Assert.Equal(expected.Length, i);
```

Comment:
Could you add testing of `GetTypeInfo` to some existing types as well?

Reply:
Yes, but not this round. I'm keeping the bug for this new API open until testing is more complete.

---
In reply to: [185133922](https://github.com/dotnet/roslyn/pull/26521#discussion_r185133922) [](ancestors = 185133922)